README.md

```diff
@@ -1,16 +1,262 @@
+
+I could preload the service with `n` UCANs with narrow time windows and `nbf`s in the future. That would certainly be very secure, but it would be less convenient since I need to come online to issue more or to 
+
+# 2 Roles
+
+Invocation adds two new roles to UCAN: invoker and executor. The existing UCAN delegator and delgate pricipals MUST persist to the invocation.
+
+| UCAN Field | Delegation                  | Invocation              |
+|------------|-----------------------------|-------------------------|
+| `iss`      | Transfer authority (active) | Request action (active) |
+| `aud`      | Receive authority (passive) | Perform action (active) |
+
+## 2.1 Invoker
+
+The invoker signals to the executor that a task associated with a UCAN SHOULD be performed.
+
+The invoker MUST be the UCAN delegator. Their DID MUST be authenticated in the `iss` field of the contained UCAN.
+
+## 2.2 Executor
+
+The executor is directed to perform some task described in the UCAN by the invoker.
+
+The executor MUST be the UCAN delegate. Their DID MUST be set the in `aud` field of the contained UCAN.
+
+# 3 Envelope
+
+The invocation envelope is a thin wrapper around a UCAN that conveys that all of the contained UCAN actions SHOULD be performed.
+
+All of the roles from the referenced UCANs MUST persist to the invocation per the [above table](FIXME).
+
+The invocation wrapper MUST be signed by the same principal that issued the UCAN.
+
+## 3.1 Fields
+
+| Field         | Type     | Value     | Descrption                                       | Required |
+|---------------|----------|-----------|--------------------------------------------------|----------|
+| `ucan/invoke` | `CID`    |           | The CID of the UCAN to invoke                    | Yes      |
+| `v`           | `SemVer` | `"0.1.0"` | SemVer of the UCAN invocation object schema      | Yes      |
+| `nnc`         | `String` |           | A unique nonce, to distinguish each invocation   | Yes      |
+| `sig`         | `Bytes`  |           | Signature of the rest of the field canonicalized | Yes      |
+  
+### 3.1.1 Invocation
+
+The `ucan/invoke` field MUST contain CIDs pointing to the UCANs to invoke. The outmost UCAN being invoked MUST NOT contain any actions that are not intended to be executed.
+
+### 3.1.2 Version
+
+The `v` field MUST contain the version of the invocation object  schema.
+
+### 3.1.3 Nonce
+
+The `nnc` field MUS contain a unique nonce. This field makes each invocation unique.
+
+### 3.1.4 Signature
+
+The `sig` field MUST contain the signature of the other fields. To construct the payload, the other fields MUST first be canonicalized as `dag-cbor`.
```

Comment:
@Gozala does this make sense, or is there a IPLD format that's especially well suited for signatures?

Reply:
Sounds good to me, but we probably should use varsig here as well

Reply:
Actually should it just say encoded as dag-cbor instead (which in turn implies canonicalization)

Reply:
I have just realized that we really need to have another tag inside the varsig to signal encoding of the signed payload. We've discussed this in https://github.com/ucan-wg/ucan-cacao/issues/2#issuecomment-1324795275 in the context of distinguishing if we're signing JWT or a CBOR. Now reading this I was thinking what if I want to sign DAG-JSON as opposed to DAG-CBOR.

I think adding another tag into varsig would be a really good idea as it would signal how data was formatted before it was signed. And in some contexts it would be necessary to allow different data encodings

Reply:
Moving this into https://github.com/ChainAgnostic/varsig/issues/3